### PR TITLE
Pulling DI back from per-request injection.

### PR DESCRIPTION
**Pull request by nyarly**
It was proving infeasible to continue to chase DI problems in the server.
This PR introduces a `ServerContext` struct that is injected as the server boots,
and then all per-request work is handled by traditional constructors.
Dependencies are still injected into the request handlers
(i.e. the `Exchanger`s)
but it's done by their Resource, pulling values from the `ServerContext.`

As a DI change, the scope of this PR is rather broad.

### CODE REVIEW

**Comment by sous-machine-account** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR: 722 lines of code changed</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: util/restful/server_test.go [hunk 1]
```diff
@@ -105,7 +122,7 @@ func TestRenderDataCanaries(t *testing.T) {
 	assert.NotZero(t, etag)
 
 	assert.Contains(t, dump, etag)
-	assert.Equal(t, dump[etag].(string), "canary")
+	assert.Equal(t, "canary", dump[etag].(string))
```
Comment (nyarly) [inline util/restful/server_test.go]: The flaw of every "assert()" - which is expected and which actual?

File: graph/graph_test.go [hunk 2]
```diff
@@ -35,6 +35,19 @@ func injectedStateManager(t *testing.T, cfg *config.Config) *StateManager {
 	g.Add(logging.SilentLogSet())
 	g.Add(newStateManager)
 	g.Add(LocalSousConfig{Config: cfg})
+	g.Add(newServerContext)
```
Comment (nyarly) [inline graph/graph_test.go]: This change makes these values actually used by this test. I'd mock them out, but this is the test that the DI system actually works...

File: server/handle_servers.go [hunk 3]
```diff
@@ -4,14 +4,17 @@ import (
...
+	"github.com/julienschmidt/httprouter"
 	"github.com/opentable/sous/config"
 	"github.com/opentable/sous/util/logging"
 	"github.com/opentable/sous/util/restful"
 )
 
 type (
 	// ServerListResource dispatches /servers
-	ServerListResource struct{}
+	ServerListResource struct {
```
[the remark this answers is not in the document]
Comment (nyarly) [reply server/handle_servers.go]: Server still means different things in this context, Hound.

**Comment by sous-machine-account** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR: 771 lines of code changed</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: util/restful/server_test.go (hunk 1 above)
Comment (stevenschlansker) [reply util/restful/server_test.go]: Not sure if Go has anything similar, but we're very happy with [assertj](http://assertj.org] which solves the problem with fluent syntax.

**Pull request merged**